bugsnag/src/test/java/com/bugsnag/ThreadMetaDataTest.java

```diff
@@ -0,0 +1,185 @@
+package com.bugsnag;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+
+import com.bugsnag.delivery.Delivery;
+
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+
+import java.util.Map;
+
+public class ThreadMetaDataTest {
+
+    private static StubNotificationDelivery delivery;
+    private static Delivery originalDelivery;
+    private static Bugsnag bugsnag;
+
+    /**
+     * Create a new test delivery and assign it to the Bugsnag client
+     */
+    @Before
+    public void swapDelivery() {
+        bugsnag = Bugsnag.init("apikey");
+        originalDelivery = bugsnag.getDelivery();
+        delivery = new StubNotificationDelivery();
+        bugsnag.setDelivery(delivery);
+    }
+
+    /**
+     * Restore the previous delivery objects after the test
+     */
+    @After
+    public void revertDelivery() {
+        bugsnag.setDelivery(originalDelivery);
+    }
+
+    @Test
+    public void testMetaDataRemoval() {
+
+        // Add some thread meta data
+        Bugsnag.addThreadMetaData("thread", "some key", "some thread value");
+
+        bugsnag.notify(new RuntimeException("test"));
+        Bugsnag.clearThreadMetaData();
+        bugsnag.notify(new RuntimeException("test"));
+
+        // Check that two reports were sent to Bugsnag
+        assertEquals(2, delivery.getNotifications().size());
+
+        // Check the meta data is added to the first report
+        Notification notification = delivery.getNotifications().get(0);
+        Report report = notification.getEvents().get(0);
+        assertTrue(report.getMetaData().containsKey("thread"));
+        assertEquals("some thread value", getMetaDataMap(notification, "thread").get("some key"));
+
+        // Check the meta data is not added to the second report
+        notification = delivery.getNotifications().get(1);
+        report = notification.getEvents().get(0);
+        assertFalse(report.getMetaData().containsKey("thread"));
+    }
+
+    @Test
+    public void testInnerThreadMetaData() {
+
+        // Add some thread meta data in the outer thread
+        Bugsnag.addThreadMetaData("outerthread", "some key", "value should not be in report");
+
+        Thread thread = new Thread(new Runnable() {
+            @Override
+            public void run() {
+                // Add thread meta data which should get associated with the exception
+                Bugsnag.addThreadMetaData("innerthread", "some key", "value should be in report");
+
+                // Notify to Bugsnag
+                bugsnag.notify(new RuntimeException("test"));
+            }
+        });
+        thread.start();
+
+        // Wait for thread to run
+        try {
+            Thread.sleep(2000);
```

Comment:
The tests in this file currently take 6 seconds to run due to the use of `Thread.sleep`. We should rework these tests to use proper synchronisation mechanisms - for example, a `CountDownLatch`.

Reply:
changed to use `thread.join();`